### PR TITLE
Retrieving and updating the location 

### PR DESCRIPTION
**Pull request by MariemBaccari**
This PR implements the location feature by:
- Adding a location helper and a location service classes 
- Asking permission for the coarse location when the user accesses the explore fragment 
- Updating the database with the location 
- A refresh button to compensate the database and location requests delays because the posts can be created before everything is updated

Note: If the user doesn't have a valid location on the database (because they didn't allow it yet), it will display the reviews from all the users on the database that the user isn't following instead of the nearby ones.

### CODE REVIEW

File: app/src/androidTest/java/com/github/sdp/mediato/ExploreFragmentTest.java [hunk 1]
```diff
@@ -95,25 +109,24 @@ public static void cleanDatabase() {
...
     @Test
     public void testItemCount() throws InterruptedException {
         Thread.sleep(5000);
         ViewInteraction outerRecyclerView = onView(withId(R.id.explore_posts));
-        //assertRecyclerViewItemCount(R.id.explore_posts, 2);
-        //outerRecyclerView.check(matches(hasItemCount(2)));
-        UserDatabase.database.getReference().setValue(null);
-        CollectionsDatabase.database.getReference().setValue(null);
-    }*/
+        assertRecyclerViewItemCount(R.id.explore_posts, 2);
```
Comment (simone-kalbermatter) [inline app/src/androidTest/java/com/github/sdp/mediato/ExploreFragmentTest.java:128]: You could add a comment explaining where the 2 comes from here. Also, it would be better to define `ViewInteraction outerRecyclerView` at the beginning of the class, the same as the other view interactions. Since it's not a recycler view of recycler views here, I would also change the name to `exploreRecyclerView` or something.

File: app/src/main/java/com/github/sdp/mediato/data/DatabaseUtils.java [hunk 2]
```diff
@@ -29,7 +31,7 @@ class DatabaseUtils {
 
     static final int PROFILE_PIC_MAX_SIZE = 1024 * 1024; //1 Megabyte
 
-    static final int DEFAULT_RADIUS = 100;
+    public static final int DEFAULT_RADIUS = 100;
```
Comment (simone-kalbermatter) [inline app/src/main/java/com/github/sdp/mediato/data/DatabaseUtils.java:34]: You could add a comment explaining what this radius is used for and in which unit it is.

File: app/src/main/java/com/github/sdp/mediato/location/LocationHelper.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+import android.content.Intent;
+import android.content.pm.PackageManager;
+
+import androidx.activity.result.ActivityResultLauncher;
+import androidx.core.content.ContextCompat;
+
+public class LocationHelper {
+    public static String USERNAME;
+
+    public static void startTrackingLocation(Context context, Activity activity, ActivityResultLauncher<String> requestPermissionLauncher, String username) {
```
Comment (simone-kalbermatter) [inline app/src/main/java/com/github/sdp/mediato/location/LocationHelper.java:16]: Because this method is public it would be good to add a full method comment.

File: app/src/main/java/com/github/sdp/mediato/data/UserDatabase.java [hunk 4]
```diff
@@ -305,9 +305,12 @@ public static CompletableFuture<List<String>> getNearbyUsernames(String username
         List<String> nearbyUsers = new ArrayList<>();
         getSavedLocation(username).thenAccept(
                 location -> {
-                    if (!location.isValid())
-                        future.completeExceptionally(new Exception("we don't have the current user's location on the database"));
+                    if (!location.isValid()) {
+                        System.out.println("No location, fetching all usernames");
```
Comment (simone-kalbermatter) [inline app/src/main/java/com/github/sdp/mediato/data/UserDatabase.java:309]: It would be better to use the logger here instead of print out.

**Review by simone-kalbermatter — CHANGES_REQUESTED** [review]
Once I enable location services and clicked refresh my explore screen is empty, but I think that is because the users I follow are too far away, right? (I am not in Lausanne :) ) Looks like impressive work implementing the location feature! Just a few comments.

File: app/src/androidTest/java/com/github/sdp/mediato/ExploreFragmentTest.java (hunk 1 above)
Comment (MariemBaccari) [reply app/src/androidTest/java/com/github/sdp/mediato/ExploreFragmentTest.java:128]: The outerRecyclerView value is actually not used so I just removed it there.

File: app/src/main/java/com/github/sdp/mediato/data/DatabaseUtils.java (hunk 2 above)
Comment (MariemBaccari) [reply app/src/main/java/com/github/sdp/mediato/data/DatabaseUtils.java]: Done

File: app/src/main/java/com/github/sdp/mediato/location/LocationHelper.java (hunk 3 above)
Comment (MariemBaccari) [reply app/src/main/java/com/github/sdp/mediato/location/LocationHelper.java:23]: Done

File: app/src/main/java/com/github/sdp/mediato/data/UserDatabase.java (hunk 4 above)
Comment (MariemBaccari) [reply app/src/main/java/com/github/sdp/mediato/data/UserDatabase.java]: Done

**Comment by MariemBaccari** [comment]
@simone-kalbermatter I just checked and the default radius is 100 km which means that you were a little bit too far indeed :) I increased it and tested it and I see your reviews now. I put it to 250 km. Also, thanks !

File: app/src/main/java/com/github/sdp/mediato/ui/viewmodel/ExploreViewModel.java [hunk 5]
```diff
@@ -34,7 +35,7 @@ public ExploreViewModel(@NonNull Application application) {
      * Sets the username of the user who is currently logged in and generates the list of posts
      * @param username the username of the user who is currently logged in
      */
-    public void setUsername(String username){
+    public void setData(String username) {
         this.username = username;
         createNearbyUsersPosts();
     }
```
Comment (sgueissa) [inline app/src/main/java/com/github/sdp/mediato/ui/viewmodel/ExploreViewModel.java:41]: Why did you change the name of the function with Data and not Username when this function is to set the username?

File: app/src/main/java/com/github/sdp/mediato/location/LocationService.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+import com.github.sdp.mediato.R;
+import com.github.sdp.mediato.data.UserDatabase;
+import com.google.android.gms.location.LocationCallback;
+import com.google.android.gms.location.LocationRequest;
+import com.google.android.gms.location.LocationResult;
+import com.google.android.gms.location.LocationServices;
+import com.google.android.gms.location.Priority;
+
+public class LocationService extends Service {
+    public static String ACTION_START_LOCATION_SERVICE = "startLocationService";
```
Comment (sgueissa) [inline app/src/main/java/com/github/sdp/mediato/location/LocationService.java:29]: Can you make it final?

File: app/src/main/java/com/github/sdp/mediato/location/LocationService.java [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    private void startLocationService() {
+        LocationRequest locationRequest = new LocationRequest.Builder(100)
+                .setPriority(Priority.PRIORITY_HIGH_ACCURACY)
+                .setIntervalMillis(100)
+                .setMinUpdateIntervalMillis(0)
+                .build();
+
+        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
```
Comment (sgueissa) [inline app/src/main/java/com/github/sdp/mediato/location/LocationService.java:57]: Could you also add the check for ACCESS_FINE_LOCATION?

**Review by sgueissa — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/github/sdp/mediato/location/LocationService.java (hunk 7 above)
Comment (MariemBaccari) [reply app/src/main/java/com/github/sdp/mediato/location/LocationService.java:57]: We don't need fine location for our app

File: app/src/main/java/com/github/sdp/mediato/location/LocationService.java (hunk 6 above)
Comment (MariemBaccari) [reply app/src/main/java/com/github/sdp/mediato/location/LocationService.java]: done in latest commit

File: app/src/main/java/com/github/sdp/mediato/ui/viewmodel/ExploreViewModel.java (hunk 5 above)
Comment (MariemBaccari) [reply app/src/main/java/com/github/sdp/mediato/ui/viewmodel/ExploreViewModel.java]: I tried to pass it other data at some point then I forgot to change it back to just username. I fixed that.

**Review by sgueissa — APPROVED** [review]
LGTM now
